### PR TITLE
remove use of .data in torch/autograd

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30338 remove easy .data in torch/nn
* #30337 remote .dat from docs in torch/nn
* #30336 remove .data for torch/jit
* **#30335 remove use of .data in torch/autograd**
* #30334 remove dead python code that uses .data
* #30258 Remove use of .data in optimizers
* #30257 Make all optimizers consistent so that they don't change gradients inplace

Differential Revision: [D18665458](https://our.internmc.facebook.com/intern/diff/D18665458)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -98,7 +97,7 @@ def get_stride(size):
             if len(input) != 1:
                 raise ValueError('gradcheck currently only supports functions with 1 input, but got: ',
                                  len(input))
-            x_tensor = x_tensor.data
+            x_tensor = x_tensor
```
Comment (vincentqb) [inline torch/autograd/gradcheck.py]: This line can just be removed.

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -116,7 +115,7 @@ def get_stride(size):
                 r = (outb - outa) / (2 * eps)
                 d_tensor[d_idx] = r.detach().reshape(-1)
         else:
-            x_tensor = x_tensor.data
+            x_tensor = x_tensor
```
Comment (vincentqb) [inline torch/autograd/gradcheck.py]: nit: same.

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -3,13 +3,14 @@
 import torch.testing
 from itertools import product
 import warnings
+from .grad_mode import no_grad
```
Comment (vincentqb) [inline torch/autograd/gradcheck.py]: Too bad this can't do `from torch.autograd import no_grad` as done in the other places. :)

Comment (albanD) [reply torch/autograd/gradcheck.py]: In the `autograd` `__init__` file, the line that adds no_grad is just below the one that adds gradcheck :/

**Review by vincentqb — CHANGES_REQUESTED** [review]
There are lots of test errors, e.g. [here](https://circleci.com/gh/pytorch/pytorch/3705937?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) :)

**Comment by albanD** [comment]
After investigation, the problem is the same as: https://github.com/pytorch/pytorch/pull/13452#issuecomment-439232114 
The Broadcast Function from distributed does bad things.
This PR fix the inplace detection for different models, and the Broadcast Function should definitely not be doing what it does.

**Comment by albanD** [comment]
Closing this until we fix the Broadcast issue described in  https://github.com/pytorch/pytorch/pull/13452#issuecomment-439232114
